keps/sig-apps/20190715-rollingUpdate-enhancement-for-daemonset.md

```diff
@@ -0,0 +1,179 @@
+---
+title: RollingUpdate enhancement for Daemonset
+authors:
+  - "@resouer"
+  - "@zhangxiaoyu-zidif"
+  - "@answer1991"
+owning-sig: sig-apps
+participating-sigs:
+  - sig-apps
+reviewers:
+  - "@janetkuo"
+approvers:
+  - TBD
+editor: TBD
+creation-date: 2019-07-15
+last-updated: 2019-07-15
+status: provisional
+see-also:
+  - n/a
+replaces:
+  - n/a
+superseded-by:
+  - n/a
+---
+
+# RollingUpdate enhancement for Daemonset
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+    - [Story 1](#story-1)
+  - [Implementation Details](#implementation-details)
+    - [API Changes](#api-changes)
+    - [Implementation](#implementation)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [Upgrades/Downgrades](#upgradesdowngrades)
+  - [Tests](#tests)
+- [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+- [Drawbacks [optional]](#drawbacks-optional)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+The purpose of this enhancement is to enhance RollingUpdate for Daemonset.
+  - When Daemonset rolling update, it can be `Paused` by user.
+  - Add `SurgingRollingUpdate` for Daemonset as a new update strategy.
+  - Add two more fields, `partition` and `selector` for `RollingUpdate` and `SurgingRollingUpdate`.
+    - `partition` is the number of DaemonSet pods remained to be old version.
+    - `selector` is to query over nodes whoes labels are matched by the Daemonset `RollingUpdate` or `SurgingRollingUpdate`.
+
+
+## Motivation
+
+Consider the following scenarios:-
+
+1. When executing Daemonset update, even those Pods run well, user need to pause updation to check if  some results of new version image meet expect. If not, user can rollback images to stable version.
+
+1. In some case, user do not want to stop some services when rolling update, so SurgingRollingUpdate is a must choice.
```

Comment:
Would you clarify this more? This sentence is a bit vague. Try making it more concrete/convincing. Maybe include some real world examples. 

Also note that DaemonSet rolling update is designed to be delete-first-create-later (and only create when the old Pod is gone) because most DaemonSet pods cannot tolerate two instances running in the same node at a time. For example, it's common to have `hostPort` specified - how do you handle that? 

Ref a previous proposal for the same effort: https://github.com/kubernetes/community/pull/977

Reply:
OK, I would give more real examples soon.

Reply:
Now In our cluster, we deploy DNS resolver and other containers serving infrastructure by daemonset in every nodes. As an online e-commercial services which have 200K+ QPS, we can not stops those services even in a very short time. So surgingRollingUpdate is a necessary strategy for our production environment.

Reply:
PS. Adding partition and Selector for SurgingRollingUpdate will help user to find some errors such as hostPort conflicts, in a limit range of 404 disaster.